File: A1/IntegerBag/IntegerBag/Menu.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace IntegerBag
{
    public class Menu
    {
        #region  Attributes
        private IntegerBag bag = new IntegerBag();
        #endregion

        #region Constructors
        public Menu() { }
        #endregion

        #region Methods
        public void Run()
        {
            int n = 0;
            do
            {
                PrintMenu();
                try
                {
                    n = int.Parse(Console.ReadLine());
                }
                catch (System.FormatException)
                {
                    Console.Clear();
                    n = -1;
                    Console.WriteLine("Please enter a valid input!!");
                    enterToContinue();
                }
                switch (n)
                {
                    case 0:
                        break;
                    case 1:
                        addElement();
                        break;
                    case 2:
                        removeElement();
                        break;
                    case 3:
                        getFrequency();
                        break;
                    case 4:
                        mostFrequent();
                        break;
                    case 5:
                        printBag();
                        break;
                    default:
                        Console.Clear();
                        Console.WriteLine("Please enter a number between 1 and 5:");
                        enterToContinue();
                        break;
                }
            } while (n != 0);
            Console.Clear();
            Console.WriteLine("Thanks for using IntegerBag");
            
        }

        static private void PrintMenu()
        {
            Console.WriteLine("Welcome to IntegerBag");
            Console.WriteLine("Please feel free to use these operations with your bag");
            Console.WriteLine("\n\n 0. - Quit the Program");
            Console.WriteLine(" 1. - Add an element to the Bag");
            Console.WriteLine(" 2. - Remove an element from the Bag");
            Console.WriteLine(" 3. - Get the frequency of an element");
            Console.WriteLine(" 4. - Get the most frequent element of the Bag");
            Console.WriteLine(" 5. - Print your current Bag");
            Console.Write("\nPlease write a number to Start: ");
        }
        private void addElement()
        {
            Console.Clear();
            int element = 0;
            bool success;
            do
            {
                success = false;
                Console.WriteLine("Please write an Integer: ");
                try
                {
                    element = int.Parse(Console.ReadLine());
                    bag.insertInt(element);
                    success = true;
                }
                catch (System.FormatException)
                {
                    Console.Clear();
                    success = false;
                    Console.WriteLine("Please enter a valid Integer!!");
                }

            } while (!success);
            Console.WriteLine("Element {0} was added to the Bag", element);
            enterToContinue();
        }
        private void removeElement()
        {
            Console.Clear();
            if (bag.isEmpty())
            {
                Console.WriteLine("This Bag has no elements yet");
                enterToContinue();
                return;
            }
            int element = 0;
            bool success;
            do
            {
                success = false;
                Console.WriteLine("Please write an element you would like to remove from the Bag");
                try
                {
                    element = int.Parse(Console.ReadLine());
                    bag.removeInt(element);
                    success = true;
                }
                catch (System.FormatException)
                {
                    Console.Clear();
                    success = false;
                    Console.WriteLine("Please enter a valid Integer!!");
                    enterToContinue();
                }
                catch (IntegerBag.NonExistingElementException)
                {
                    Console.Clear();
                    success = false;
                    Console.WriteLine("This element does not exist in this Bag");
                    enterToContinue();
                }
            } while (!success);
            if (success)
            {
                Console.WriteLine("Element {0} was removed to the Bag", element);
            }
            enterToContinue();
        }
        private void getFrequency() 
        {
            Console.Clear();
            if (bag.isEmpty())
            {
                Console.WriteLine("This Bag has no elements yet");
                enterToContinue();
                return;
            }
            int element = 0;
            int frequency = 0;
            bool success;
            do
            {
                success = false;
                Console.WriteLine("Please write an element to get its frequency");
                try
                {
                    element = int.Parse(Console.ReadLine());
                    frequency = bag.frequencyOf(element);
                    success = true;
                }
                catch (System.FormatException)
                {
                    Console.Clear();
                    success = false;
                    Console.WriteLine("Please enter a valid Integer!!");
                    enterToContinue();
                }
                catch (IntegerBag.NonExistingElementException)
                {
                    Console.Clear();
                    success = false;
                    Console.WriteLine("This element does not exist in this Bag");
                    enterToContinue();
                }
            } while (!success);
            if (success)
            {
                Console.WriteLine("Element {0} frequency is {1}", element, frequency);
            }
            enterToContinue();
        }
        private void mostFrequent()
        {
            Console.Clear();
            int maxIndex = 0;
            int element = 0;
            int frequency = 0;
            bool success;
            do
            {
                try
                {
                    maxIndex = bag.mostFrequent(out element, out frequency);
                    success = true;
                }
                catch (System.FormatException)
                {
                    Console.Clear();
                    success = false;
                    Console.WriteLine("Please enter a valid Integer!!");
                    enterToContinue();
                }
                catch (IntegerBag.NonExistingElementException)
                {
                    Console.Clear();
                    success = false;
                    Console.WriteLine("This element does not exist in this Bag");
                    enterToContinue();
                }
                catch (IntegerBag.EmptyBagException)
                {
                    Console.Clear();
                    success = false;
                    Console.WriteLine("This Bag has no elements yet");
                    break;
                }

            } while (!success);
            if (success)
            {
                Console.WriteLine("Element {0} is the most frequent element, its frequency is {1}", element, frequency);
            }
            enterToContinue();
        }
        private void printBag()
        {
            Console.Clear();
            Console.WriteLine("The bag structure is: [ (element, frequency)]");
            Console.WriteLine("This is your Bag: ");
            Console.WriteLine(bag.ToString());
            enterToContinue();
        }
        #endregion

        #region Complementary Methods
        static private void enterToContinue()
        {
            Console.WriteLine("Press enter to continue");
            Console.ReadLine();
            Console.Clear();
        }
        #endregion
    }
}

File: A1_py/Menu.py
import IntegerBag
import os


class menu:
    def __init__(self):
        self.__bag = IntegerBag.IntegerBag()

    def printMenu(self):
        print(
            """
Hello Welcome to the Integer Bag Program, please enter an option:

1) Add an element to the Bag
2) Remove an element from the Bag
3) Get frequency of an element from the Bag 
4) Get the most frequent element from the bag
5) Print the elements of the Bag

Enter 'exit' to leave the program""") 

    def pressContinue(self):
        input("Press Enter to continue:\n")
        os.system('clear')

    def Run(self):
        option = None
        while True:
            self.printMenu()
            option = input("Please enter a number:\n")
            try:
                option = int(option)
            except ValueError:
                if option == "exit":
                    os.system('clear')
                    print("See you next time!")
                    break
                else:
                    os.system('clear')
                    print("Please enter a number")
                    option = -1
                    self.pressContinue()
            os.system('clear')
            match option:
                case 1:
                    self.__Add() 
                case 2:
                    self.__Remove()
                case 3:
                    self.__Frequency()
                case 4:
                    self.__MostFrequent()
                case 5:
                    self.__Print()  
                case _:
                    if option != -1:
                        print("Please enter a valid number")
                        self.pressContinue()

    def __Add(self):
        while True:
            try:
                sucess = True
                element =int(input("Please enter a number to add to the Bag:\n"))
                self.__bag.Insert(element)
            except ValueError:
                os.system('clear')
                print("Please enter a number")
                sucess = False
                self.pressContinue()
            except Exception as e:
                os.system('clear')
                print(str(e))
                sucess = False
                self.pressContinue()
            if sucess:
                print(f"The number {element} was added to the bag successfully")
                self.pressContinue()
                break

    def __Remove(self):
            if self.__bag.Length() == 0:
                print(f"The bag is empty, please enter a number to the bag first!")
                self.pressContinue()
            else:
                while True:
                    try:
                        sucess = True
                        print(f"This is the bag {self.__bag}")
                        element = int(input("Please enter a number to add to the Bag:\n"))
                        self.__bag.Remove(element)
                    except ValueError:
                        os.system('clear')
                        print("Please enter a number")
                        sucess = False
                        self.pressContinue()
                    except IntegerBag.ElementNotInBagException:
                        os.system('clear')
                        print("This number is not in the bag, please enter a number from the bag")
                        sucess = False
                        self.pressContinue()
                    except Exception as e:
                        os.system('clear')
                        print(str(e))
                        sucess = False
                        self.pressContinue()
                    if sucess:
                        print(f"The number {element} was removed to the bag successfully")
                        self.pressContinue()
                        break
                    
    def __Frequency(self):
        if self.__bag.Length() == 0:
            print("The bag is empty, please enter a number to the bag first!")
            self.pressContinue()
        else:
            while True:
                try:
                    sucess = True
                    print(f"This is the bag {self.__bag}")
                    element =  int(input("Please enter a number to obtain the frequency:\n"))
                    freq = self.__bag.GetFrequency(element)
                except ValueError:
                    os.system('clear')
                    print("Please enter a number")
                    sucess = False
                    self.pressContinue()
                except IntegerBag.ElementNotInBagException:
                    os.system('clear')
                    print("This number is not in the bag, please enter a number from the bag")
                    sucess = False
                    self.pressContinue()
                except Exception as e:
                    os.system('clear')
                    print(str(e))
                    sucess = False
                    self.pressContinue()
                if sucess:
                    print(f"The number {element} has a frequency of {freq}")
                    self.pressContinue()
                    break

    def __MostFrequent(self):
        if self.__bag.Length() == 0:
            print("The bag is empty, please enter a number to the bag first!")
            self.pressContinue()
        else:
            return self.__bag.GetMax()

    def __Print(self):
            print(f"This is your bag:\n{self.__bag}")
